pyclam/manifold.py

```diff
@@ -849,6 +822,46 @@ def clear_cache(self):
         self.cache = dict()
         return
 
+    @property
+    def ordered_clusters(self) -> List[Cluster]:
+        if 'ordered_clusters' not in self.cache:
+            self.cache['ordered_clusters'] = [
+                cluster for layer in self.layers
+                for cluster in layer.clusters
+                if cluster.depth == layer.depth
+            ]
+        return self.cache['ordered_clusters']
+
+    @property
+    def ratios(self) -> Dict[Cluster, np.array]:
+        if 'ratios' not in self.cache:
+            smoothing, period = 2, 10
+            alpha = smoothing / (1 + period)
+            cluster_indices: Dict[Cluster, int] = {cluster: i for i, cluster in enumerate(self.ordered_clusters)}
+            cluster_ratios: np.array = np.ones(shape=(len(self.ordered_clusters), 6), dtype=float)
+            for layer in self.layers[1:]:
+                for cluster in layer.clusters:
+                    current = np.asarray([
+                        cluster.cardinality / cluster.parent.cardinality,
+                        cluster.radius / (cluster.parent.radius + EPSILON),
+                        cluster.local_fractal_dimension / cluster.parent.local_fractal_dimension,
+                    ], dtype=float)
+                    previous = cluster_ratios[cluster_indices[cluster.parent]][3:]
+                    new = alpha * current + (1 - alpha) * previous
+                    cluster_ratios[cluster_indices[cluster]] = np.concatenate([current, new])
+
+            for i in range(cluster_ratios.shape[1]):
+                cluster_ratios[:, i] = normalize(cluster_ratios[:, i], 'gaussian')
```

Comment:
do we *always* want to normalize eagerly? should this be a toggle?

Reply:
chaoda performance was always worse when the ratios were not normalized. I'm open for adding a toggle for normalization but it would complicate caching. Once we figure out that complexity, we can add it